### ISSUE TITLE
Use with Quarto?

### ISSUE THREAD

**Issue by FeralFlora** [ram02z/d2-filter#6]
Thanks for making this filter for D2! I was hoping to use it with [Quarto](https://quarto.org/), but it doesn't seem possible to do so as is.

Therefore, I wanted to ask if you knew more about how this could be achieved, or if you would be interested in making a Quarto distribution of this filter. See: 
- https://quarto.org/docs/extensions/creating.html
- https://quarto.org/docs/extensions/distributing.html

**Comment by ram02z** [comment]
Hi, I actually tried out Quarto a few weeks back, and was impressed with the results.

From what I understand, Quarto does support pandoc filters, so I am not sure why this filter doesn't work out of the box.

Thank you for flagging this, I'll look into this further this weekend.

**Comment by FeralFlora** [comment]
> was impressed with the results.

Yeah, me too, it's very nice.

> Quarto does support pandoc filters, so I am not sure why this filter doesn't work out of the box.

It does, but they are specified in the _quarto.yml file or the YAML at other levels (Project sub-directories, or in the document itself). They are not called as command arguments directly, like you do in your example in the read me. Also, I am not sure where Quarto looks for filters, and where npm installs your D2 filter. Lastly, I am a bit confused by the .cmd ending of the filter in the example you provide for Windows. Is this not a Lua filter? Quarto recommends Lua filters or their filter extensions.

I tried installing via npm, then specifying it under `filters` in my Quarto document's YAML, as the [Quarto docs](https://quarto.org/docs/extensions/filters.html) instruct:

```
filters:
  - d2-filter.cmd
```
I tried this with and without .cmd.

>Thank you for flagging this, I'll look into this further this weekend.

No problem, and thanks for the response, looking forward to hearing more.

**Comment by FeralFlora** [comment]
@ram02z See this if you are having issues: https://github.com/quarto-dev/quarto-cli/issues/5392

**Comment by ram02z** [comment]
> where npm installs your D2 filter. 

Since you are on Windows, the filter should be accessible by running `d2-filter.cmd` if you installed the filter globally using the `-g` flag, i.e. `npm i -g d2-filter`. I should probably make this clearer in the README.

> Lastly, I am a bit confused by the .cmd ending of the filter in the example you provide for Windows. Is this not a Lua filter? Quarto recommends Lua filters or their filter extensions.

This is not a lua filter, this is a JSON filter which I created using this excellent pandoc-filter library (https://github.com/mvhenderson/pandoc-filter-node)

I briefly looked at the issue you linked and it seems like Quarto might not support JSON filters properly. However, I managed to get the filter to work using quarto, but you have to have the full path of the filter for it to work. Seems like Quarto expects the filter to be in the same directory(?)

```
---
title: Test
filters:
  - /usr/bin/d2-filter
---

~~~{.d2}
x -> y
~~~
```

You will need to replace `/usr/bin/d2-filter` with the path of the binary on your system. You should be able to run `where.exe d2-filter.cmd` to find the full path.

**Comment by ram02z** [comment]
I will close the issue for now. @MSBack you can reopen the issue if you are still having problems.

**Comment by FeralFlora** [comment]
> Seems like Quarto expects the filter to be in the same directory(?)

Yes, I think you apply them on a project by project basis.

I couldn't get it to work using your example. I reinstalled using `npm i -g d2-filter`.

Then, `where d2-filter` gives me these paths:
```
C:\Users\User\AppData\Roaming\npm\d2-filter
C:\Users\User\AppData\Roaming\npm\d2-filter.cmd
```

If I use the first path without .cmd, I get this error and the preview fails:
```
Error running filter C:/Users/User/AppData/Local/Programs/Quarto/share/filters/main.lua:
Error running filter C:\Users\User\AppData\Roaming\npm\d2-filter:
C:\Users\User\AppData\Roaming\npm\d2-filter: createProcess: invalid argument (Exec format error)
```

If I user the second path including `.cmd`, the preview works and the filter is accepted. However, the diagram is not rendered.

Here's my test diagram:
```
~~~{.d2}
a -> b
~~~
```

Note that I am previewing as Reveal.js.

**Comment by FeralFlora** [comment]
Update: It works when previewing as HTML. Preview as PDF or Reveal.js only gives me a caption (If I add that).

**Comment by FeralFlora** [comment]
Btw, I cannot re-open the issue, even though I want to, because I was not the one who closed it.

**Comment by elipousson** [comment]
Just a heads up that I shared this filter to [a related discussion on the Quarto repository](https://github.com/quarto-dev/quarto-cli/discussions/4002) and @cderv put together a prototype Quarto extension that wraps this filter: https://github.com/cderv/run-d2-filter

I don't think I have d2 and d2-filter on PATH in the way required to get the extension working but it seems like great progress towards this integration.

**Comment by cderv** [comment]
Sorry I missed this issue. I indeed did a quick prototype by wrapping inside Lua. 

Quarto works well with Lua filter (as we use it extensively). It should work also with JSON filter, but path is not resolved as discussed in
* https://github.com/quarto-dev/quarto-cli/issues/5392

It was also to have a way to resolve the `.cmd` extension issue with Pandoc. 

Maybe wrapping all the necessary JS for this to work in a `_extension` would make path findable by Quarto when `filters` is a JSON one. I did not try yet.  

Happy to help in anyway
